### PR TITLE
GTIR domain inference

### PR DESCRIPTION
**Pull request by SF-N**
#### Uses TraceShifts to infer the minimal domain of (nested) `as_fieldop`s and SetAt:
Domain inference of let-statements is not implemented yet and will be done in PR #1591
```
expr = as_fieldop(stencil)(as_fieldop(stencil)(inp_field))

domain = cartesian_domain(...)

# new_expr has transformed all as_fieldop(stencil) into as_fieldop(stencil, domain)
infer_as_fieldop(expr, domain) -> new_expr, dict(referenced_field => accessed domain)
# with new_expr = as_fieldop(stencil, cartesian_domain(...))(inp_field, ...)

# multiple SetAt (no composition)
tmp <- as_fieldop(stencil)(inp_field, ...) @ auto_domain()
out <- as_fieldop(stencil)(tmp, ...) @ ...

# domain is inferred bottom up from last SetAt, only domains of SetAt on temporaries are changed
infer_program(program_expr)
```
### PR Description
- New file gt4py.next.iterator.transforms.infer_domain with function `infer_as_fieldop` and `infer_program`
- Several new tests in test_infer_domain.py to test functionality
- New function SymbolicDomain.translate
- New helper-function `domain` and `as_fieldop` in ir.makers
- Uses constant folding of `itir.Literal`s to simplify resulting domains

### Todos
- [x] Domain inference for nested  `as_fieldop`s (cartesian_domain)
- [x] Domain inference for SetAt
- [x] Extend to unstructured grids

### Note
I agree that my contribution in this PR is additionally licensed under the BSD-3-Clause license (SPDX short identifier).

### CODE REVIEW

File: src/gt4py/next/iterator/ir_utils/ir_makers.py [hunk 1]
```diff
@@ -389,3 +396,34 @@ def _impl(*its: itir.Expr) -> itir.FunCall:
 def map_(op):
     """Create a `map_` call."""
     return call(call("map_")(op))
+
+
+def cartesian_domain(ranges: Dict[str, Tuple[Any, Any]], out_field: str,
```
Comment (tehrengruber) [inline src/gt4py/next/iterator/ir_utils/ir_makers.py:401]: This function looks unnecessarily complicated, just a dict with start end stop of the ranges should be sufficient. This would be a good doctest to add.
```
>>> cartesian_domain({"IDim": (0, 10), "JDim": (0, 20)})
c⟨ IDim: [0, 10), JDim: [0, 20) ⟩
```

File: src/gt4py/next/iterator/ir_utils/ir_makers.py [hunk 2]
```diff
@@ -389,3 +396,34 @@ def _impl(*its: itir.Expr) -> itir.FunCall:
...
+        size_field = [None] * len(ranges)
+
+    symbolic_ranges = {}
+
+    for index, (axis, (start, stop)) in enumerate(ranges.items()):
+        if axis not in axis_order:
+            raise ValueError("The ranges need to contain either IDim, JDim or KDim.")
+
+        axis_index = axis_order.index(axis)
+        start_expr = ConstantFolding.apply(start)
```
Comment (tehrengruber) [inline src/gt4py/next/iterator/ir_utils/ir_makers.py:417]: You only need the constant folding for readability in your tests so this should happen there. The makers should not contain any calls to passes or "optimizations", but just be simple tools that ease creation of nodes.

File: src/gt4py/next/iterator/transforms/infer_domain.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+            return SymbolicDomain.from_expr(domain)
+        raise TypeError("domain must either be a FunCall or a SymbolicDomain.")
+
+    @classmethod
+    def _translate_domain(cls, symbolic_domain: SymbolicDomain, shift: Tuple[itir.OffsetLiteral, int],
+                          dims: List[str]) -> SymbolicDomain:
+        new_ranges = {dim: symbolic_domain.ranges[dim] for dim in dims}
+        if shift:
+            off, val = shift
+            current_dim = cls._infer_dimension_from_offset(off)
```
Comment (tehrengruber) [inline src/gt4py/next/iterator/transforms/infer_domain.py:72]: You will need the offset provider for connectivities (i.e. unstructured) when you extend this to not only support cartesian dimensions later on. It will be an argument to the final pass that should be forwarded here.
```suggestion
            current_dim = offset_provider[off]
            assert isinstance(current_dim, Dimension)
```
For now it would make sense to
- Add a fixture to your tests that provides you with a cartesian offset provider
  ```
  offset_provider={"IOff": Dimension(...), "JOff": Dimension(...)}
- Pass it as an argument to `infer_as_fieldop` and pass it down the chain until here.

File: src/gt4py/next/iterator/transforms/infer_domain.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+            new_ranges[current_dim.value] = SymbolicRange.translate(
+                symbolic_domain.ranges[current_dim.value], val.value
+            )
+        return SymbolicDomain("cartesian_domain", new_ranges)
+
+    @classmethod
+    def infer_as_fieldop(
+            cls,
+            stencil: itir.FunCall,
```
Comment (tehrengruber) [inline src/gt4py/next/iterator/transforms/infer_domain.py:81]: I think what you call stencil here is an `applied_fieldop`, e.g `as_fieldop(stencil, domain)(field_args...)`. A stencil is a function from Iterators to a value, e.g. `lambda x: deref(x).`

File: src/gt4py/next/iterator/transforms/infer_domain.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+            input_domain: SymbolicDomain
+    ) -> Tuple[itir.FunCall, Dict[str, SymbolicDomain]]: # todo: test scan operator
+
+        assert isinstance(stencil, itir.FunCall) and isinstance(stencil.fun, itir.FunCall)
+        assert stencil.fun.fun == im.ref('as_fieldop')
+
+        inputs = stencil.args
```
Comment (tehrengruber) [inline src/gt4py/next/iterator/transforms/infer_domain.py:88]: just for readability (stencil has a different meaning here see comment above).
```suggestion
        # `as_fieldop(stencil, domain)(inputs...)`
        stencil, inputs = applied_fieldop.fun.args[0], applied_fieldop.args
```

File: src/gt4py/next/iterator/transforms/infer_domain.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+        symbolic_domain = cls._get_symbolic_domain(input_domain)
+
+        # Build node that can be passed to TraceShifts
+        node = itir.StencilClosure(
+            stencil=stencil.fun.args[0],
+            inputs=inputs_node,
+            output=im.ref(out_field_name),
+            domain=SymbolicDomain.as_expr(symbolic_domain),
+        )
```
Comment (tehrengruber) [inline src/gt4py/next/iterator/transforms/infer_domain.py:110]:
```suggestion
# TODO: until TraceShifts directly supporty lambda functions we just wrap our expression into
  # a dummy closure in this helper function.
def trace_shifts(stencil: ir.Expr, inputs: list[ir.Expr], domain: ir.Expr):
        node = itir.StencilClosure(
            stencil=stencil,
            inputs=inputs_node,
            output=im.ref(out_field_name),
            domain=SymbolicDomain.as_expr(symbolic_domain),
        )
```

File: src/gt4py/next/iterator/transforms/infer_domain.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+                transformed_calls_tmp, accessed_domains_tmp = cls.infer_as_fieldop(in_field,
+                                                                                   accessed_domains[str(in_field.id)])
+                inputs_new[idx] = transformed_calls_tmp
+
+                # Merge accessed_domains and accessed_domains_tmp
+                for k, v in accessed_domains_tmp.items():
+                    if k in accessed_domains:
+                        accessed_domains[k] = domain_union([accessed_domains[k], v])
+                    else:
+                        accessed_domains[k] = v
```
Comment (tehrengruber) [inline src/gt4py/next/iterator/transforms/infer_domain.py:138]: I think this belongs to the beginning of this function. This will also cleanup the awkward multiple copies of the inputs (or I am overlooking something...).

File: tests/next_tests/unit_tests/iterator_tests/transforms_tests/test_domain_inference.py [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+    expected = im.call(im.call('as_fieldop')( stencil, domain))(im.ref('in_field1'),im.call(im.call('as_fieldop')(inner_stencil, domain_inner))(im.ref('in_field1'), im.ref('in_field2')))
+    expected_domains = {'in_field1': SymbolicDomain.from_expr(im.cartesian_domain({"IDim": (1, 1), "JDim": (-1, 0)}, "tmp", size_field=[im.literal_from_value(11),im.literal_from_value(7)])),
+                        'in_field2': SymbolicDomain.from_expr(im.cartesian_domain({"IDim": (0, 0), "JDim": (-2, -2)}, "tmp", size_field=[im.literal_from_value(11),im.literal_from_value(7)]))}
+
+    actual_call, actual_domains = InferDomain.infer_as_fieldop(testee, domain)
+
+    assert actual_call == expected
+    assert actual_domains == expected_domains
+
+def test_nested_stencils_n_times():
```
Comment (tehrengruber) [inline tests/next_tests/unit_tests/iterator_tests/transforms_tests/test_domain_inference.py:139]: You can parameterize this test (https://docs.pytest.org/en/7.0.x/how-to/parametrize.html) and which should make it a little simpler because you don't need `tmp, domain, expected` to be arrays.

**Review by tehrengruber — CHANGES_REQUESTED** [review]
First course round.

File: src/gt4py/next/iterator/transforms/infer_domain.py [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+            stencil: itir.FunCall,
+            input_domain: SymbolicDomain
+    ) -> Tuple[itir.FunCall, Dict[str, SymbolicDomain]]: # todo: test scan operator
+
+        assert isinstance(stencil, itir.FunCall) and isinstance(stencil.fun, itir.FunCall)
+        assert stencil.fun.fun == im.ref('as_fieldop')
+
+        inputs = stencil.args
+        tmp_counter = 0
+        inputs_node = inputs.copy()
```
Comment (tehrengruber) [inline src/gt4py/next/iterator/transforms/infer_domain.py:90]: Having multiple variables containing the inputs doesn't look necessary, but better read the rest of the comments before you try to change this.

File: src/gt4py/next/iterator/transforms/infer_domain.py [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+                stencil=stencil,
+                inputs=inputs,
+                output=im.ref(out_field_name),
+                domain=domain,
+            )
+            return TraceShifts.apply(node)
+
+        # Extract the shifts and translate the domains accordingly
+        shifts_results = trace_shifts(stencil, inputs_node, SymbolicDomain.as_expr(symbolic_domain))
+        dims = cls._extract_axis_dims(SymbolicDomain.as_expr(symbolic_domain))
```
Comment (tehrengruber) [inline src/gt4py/next/iterator/transforms/infer_domain.py:126]: Should be possible to remove the function and also absorb it into translate domain.

**Comment by tehrengruber** [comment]
Note to myself: One function for itir.Program with temporaries, one without (only the simple for loop). When writing temporary pass introduce `ProgramWithUnresolvedTemporaries` that allows `_gtmp_auto_domain` and potentially move the inference (continuing to build upon the primitive function operaton on `as_fieldop` here) there.

File: src/gt4py/next/iterator/transforms/infer_domain.py [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+
+        temporaries = [str(tmp.id) for tmp in program.declarations]
+
+        cls._validate_temporary_usage(program.body, temporaries)
+
+        for set_at in reversed(program.body):
+            assert isinstance(set_at, itir.SetAt)
+            assert isinstance(set_at.expr, itir.FunCall)
+            assert isinstance(set_at.expr.fun, itir.FunCall)
+            assert set_at.expr.fun.fun == im.ref("as_fieldop")
```
Comment (havogt) [inline src/gt4py/next/iterator/transforms/infer_domain.py:174]: I noticed that this is too restrictive for what we need. That was not clear when this project started. We now allow also pure field expressions at this level, e.g. 

```
__field_operator_testee(
  __sym_1,
  out,
  ____sym_1_size_0,
  ____sym_1_size_1,
  ____sym_1_size_2,
  __out_size_0,
  __out_size_1,
  __out_size_2
) {
  testee = λ(a) → a;
  out @ c⟨ IDimₕ: [0, __out_size_0), JDimₕ: [0, __out_size_1), KDimᵥ: [0, __out_size_2) ⟩
       ← testee(__sym_1);
}
```

File: src/gt4py/next/iterator/transforms/infer_domain.py [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+        return SymbolicDomain("cartesian_domain", new_ranges)
+
+    @classmethod
+    def infer_as_fieldop(
+        cls,
+        applied_fieldop: itir.FunCall,
+        input_domain: SymbolicDomain,
+        offset_provider: Dict[str, Dimension],
+    ) -> Tuple[itir.FunCall, Dict[str, SymbolicDomain]]:  # todo: test scan operator
+        assert isinstance(applied_fieldop, itir.FunCall) and isinstance(
```
Comment (tehrengruber) [inline src/gt4py/next/iterator/transforms/infer_domain.py:73]:
```suggestion
        if isinstance(applied_fieldop, itir.SymRef):
          accessed_domains = {applied_fieldop.id: input_domain}
          return applied_fieldop, accessed_domains
          
        assert isinstance(applied_fieldop, itir.FunCall) and isinstance(
```
This should work for your example @havogt here https://github.com/GridTools/gt4py/pull/1568/files#r1672562626. Tuples and lets are not yet supported by this pass. The latter can be solved by inlining first.

File: src/gt4py/next/iterator/ir_utils/ir_makers.py [hunk 13]
```diff
@@ -401,3 +403,64 @@ def _impl(*its: itir.Expr) -> itir.FunCall:
...
+                else itir.AxisLiteral(value=d),
+                r[0],
+                r[1],
+            )
+            for d, r in ranges.items()
+        ]
+    )
+
+
+# TODO: merge with cartesian_domain using GridType
```
Comment (tehrengruber) [inline src/gt4py/next/iterator/ir_utils/ir_makers.py:438]:
```suggestion
def domain(grid_type: common.GridType, ranges: Dict[Union[common.Dimension, str], Tuple[Any, Any]]) -> itir.FunCall:
```
I would propose this signature.

File: src/gt4py/next/iterator/transforms/global_tmps.py [hunk 14]
```diff
@@ -451,12 +452,20 @@ def from_expr(cls, node: ir.Node) -> SymbolicDomain:
...
+            return im.cartesian_domain(converted_ranges)
+        else:
+            return im.call("unstructured_domain")(
+                *[
+                    im.call("named_range")(
+                        ir.AxisLiteral(value=d.value, kind=d.kind), r.start, r.stop
+                    )
+                    for d, r in self.ranges.items()
+                ]
+            )
```
Comment (tehrengruber) [inline src/gt4py/next/iterator/transforms/global_tmps.py:468]:
```suggestion
        im.domain(self.grid_type, ...)
```

File: src/gt4py/next/iterator/transforms/global_tmps.py [hunk 15]
```diff
@@ -473,7 +482,9 @@ def domain_union(domains: list[SymbolicDomain]) -> SymbolicDomain:
             lambda current_expr, el_expr: im.call("maximum")(current_expr, el_expr),
             [domain.ranges[dim].stop for domain in domains],
         )
-        new_domain_ranges[dim] = SymbolicRange(start, stop)
+        new_domain_ranges[dim] = SymbolicRange(
```
Comment (tehrengruber) [inline src/gt4py/next/iterator/transforms/global_tmps.py:485]: Is this change required? I hope not then we shouldn't do it here to avoid creating an unnecessary dependency between the passes. If a cleaner IR is required for tests / debugging this should be done outside of the pass.

File: src/gt4py/next/iterator/transforms/infer_domain.py [hunk 16]
```diff
@@ -0,0 +1,236 @@
...
+            else:
+                raise AssertionError()
+
+        return SymbolicDomain(symbolic_domain.grid_type, new_ranges)
+
+    @classmethod
+    def infer_as_fieldop(
+        cls,
+        applied_fieldop: itir.FunCall,
+        input_domain: SymbolicDomain,
```
Comment (tehrengruber) [inline src/gt4py/next/iterator/transforms/infer_domain.py:100]:
```suggestion
        input_domain: SymbolicDomain | ir.FunCall,
```
and then below you can do the canonicalization
```
if isinstance(input_domain, ir.FunCall):
  input_domain = SymbolicDomain.from_expr(input_domain)
```
The `_get_symbolic_domain` is then not needed anymore.

File: src/gt4py/next/iterator/transforms/infer_domain.py [hunk 17]
```diff
@@ -0,0 +1,236 @@
...
+            else:
+                inputs_node.append(in_field)
+                accessed_domains[str(in_field.id)] = []
+
+        out_field_name = "tmp"  # todo: can this be derived from somewhere?
+
+        symbolic_domain = cls._get_symbolic_domain(input_domain)
+
+        # TODO: until TraceShifts directly supporty stencils we just wrap our expression into a dummy closure in this helper function.
+        def trace_shifts(stencil: itir.Expr, inputs: list[itir.Expr], domain: itir.Expr):
```
Comment (tehrengruber) [inline src/gt4py/next/iterator/transforms/infer_domain.py:130]: Let's move this outside of the function to keep the reader focused.

File: src/gt4py/next/iterator/transforms/infer_domain.py [hunk 18]
```diff
@@ -0,0 +1,236 @@
...
+from gt4py.next.common import Dimension
+from gt4py.next.iterator import ir as itir
+from gt4py.next.iterator.ir_utils import ir_makers as im
+from gt4py.next.iterator.transforms.global_tmps import SymbolicDomain, SymbolicRange, domain_union, \
+    _max_domain_sizes_by_location_type
+from gt4py.next.iterator.transforms.trace_shifts import TraceShifts
+
+
+@dataclasses.dataclass(frozen=True)
+class InferDomain:
```
Comment (tehrengruber) [inline src/gt4py/next/iterator/transforms/infer_domain.py:27]: Let's remove this class (at least for now) and make all class method free functions. There appears to be no need to have a class.

**Review by tehrengruber — CHANGES_REQUESTED** [review]

File: src/gt4py/next/iterator/transforms/global_tmps.py (hunk 15 above)
Comment (SF-N) [reply src/gt4py/next/iterator/transforms/global_tmps.py:485]: Yes, it is required to get rid of minimum/maximum in the SymbolicDomains returned by infer_as_fieldop. Should the ConstantFolding be done directly in the tests in the end or somewhere else?

Comment (SF-N) [reply src/gt4py/next/iterator/transforms/global_tmps.py]: As discussed, I removed the ConstantFolding from domain_union and do it now directly in the tests.

File: src/gt4py/next/iterator/transforms/infer_domain.py [hunk 19]
```diff
@@ -0,0 +1,225 @@
...
+            set_at.expr, accessed_domains[set_at.target.id], offset_provider
+        )
+        transformed_set_ats.insert(
+            0,
+            itir.SetAt(
+                expr=transformed_as_fieldop,
+                domain=SymbolicDomain.as_expr(accessed_domains[set_at.target.id]),
+                target=set_at.target,
+            ),
+        )
```
Comment (havogt) [inline src/gt4py/next/iterator/transforms/infer_domain.py:198]: updating the domain should only be done if the original domain is `AUTO_DOMAIN`, correct? I am wondering if we should split this pass in 2. One that propagates domains between set_at (i.e. resolves AUTO_DOMAIN) and one that just annotates `as_fieldop`s. I find it confusing that we do both in the same pass, but not sure if I am missing something.

File: src/gt4py/next/iterator/transforms/infer_domain.py [hunk 20]
```diff
@@ -0,0 +1,206 @@
...
+            new_ranges[current_dim] = SymbolicRange.translate(
+                symbolic_domain.ranges[current_dim], val.value
+            )
+        return SymbolicDomain("cartesian_domain", new_ranges)
+
+    @classmethod
+    def infer_as_fieldop(
+        cls,
+        applied_fieldop: itir.FunCall,
+        input_domain: SymbolicDomain,
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/iterator/transforms/infer_domain.py]: ignore this, it was an old comment that I didn't submit.

[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/iterator/transforms/infer_domain.py]: ah maybe it's still valid.

[the remark this answers is not in the document]
Comment (SF-N) [reply src/gt4py/next/iterator/transforms/infer_domain.py]: Thanks, I changed it.

File: src/gt4py/next/iterator/transforms/infer_domain.py (hunk 19 above)
Comment (SF-N) [reply src/gt4py/next/iterator/transforms/infer_domain.py:172]: I am also not sure if I am missing something, but I think this is only possible with some code duplication.
One option could be to add a pass that infers the domains of all `as_fieldop`s which is then only called by DaCe.
Let's discuss this next time we talk.

**Review by tehrengruber — APPROVED** [review]

File: src/gt4py/next/iterator/transforms/infer_domain.py (hunk 19 above)
Comment (tehrengruber) [reply src/gt4py/next/iterator/transforms/infer_domain.py:178]: As discussed: We discovered that running the domain propagation prior to temporary extraction should make all temporary related code in the propagation unnecessary and can hence be removed. We've placed a todo (in the let branch) and kept it in here for now so that we can move on.

File: src/gt4py/next/iterator/transforms/infer_domain.py (hunk 12 above)
Comment (tehrengruber) [reply src/gt4py/next/iterator/transforms/infer_domain.py]: Implemented (in the let branch).

File: src/gt4py/next/iterator/transforms/infer_domain.py (hunk 11 above)
Comment (tehrengruber) [reply src/gt4py/next/iterator/transforms/infer_domain.py]: Implemented (in the let branch).
